Determinant and Adjugate Matrix

FranziskusWiesnet opened the pull request:
In the four new files, the determinant of a matrix over a commutative ring is introduced through Laplace expansion, and it is proven that the adjugate matrix multiplied by the matrix itself equals the determinant times the identity matrix.
Additionally, it is shown that the determinant is independent of the row or column chosen for expansion, along with a few other minor properties of the determinant.
I am happy to consider any improvements.

Commit:
Determinat added

Commit:
Comments added

Commit:
whitespace fix

felixwellen commented on dist-newstyle/cache/compiler:
This looks like accidentally committed files

FranziskusWiesnet replied:
Yes. I have removed it.

felixwellen commented:
Thanks for opening a PR on this! Determinants would certainly be very useful to have.

felixwellen commented on Cubical/Algebra/Determinat/Minor.agda:
```diff
@@ -0,0 +1,194 @@
+{-# OPTIONS --cubical #-}
+
+module Minor where
```
Module names should always be fully qualified, i.e. ```Cubical.Algebra.Determinant.Minor``` in this case.

Commit:
names fixed

Commit:
save-flags added

anshwad10 commented on Cubical/Algebra/Determinat/Base.agda:
```diff
@@ -0,0 +1,2500 @@
+  CommRingR' : CommRingStr (R' .fst)
+  CommRingR' =  commringstr 0r 1r _+_ _·_ -_ (CommRingStr.isCommRing (snd P'))
+
+  -- Definition of the minor factor
```
I noticed that there is some redundancy in the Determinat (is that a typo?) .Base file. `MF i` is just `(- 1r) ^ i` (`_^_` is defined in Cubical.Algebra.CommRing.Properties, and many of the properties of minor factor you prove are already proved there). And `+Compat` can by proved more simply by `cong₂ _+_`.

mortberg replied:
Sounds right. Can you fix?

felixwellen committed:
Revert "Auxiliary commit to revert individual files from 1ea2f8f0f4a9c000fafb882e5b6046408f6993e0"

This reverts commit e7aee8949ba8708c0ffbdb8f8d21fbdb13af6153.

felixwellen committed:
use opaque, don't duplicate --cubical, order imports, use varibales

FranziskusWiesnet committed:
Merge pull request #1 from agda/fwellen/fixes-determinant

Some fixes for your PR on determinants

mortberg commented on Cubical/Algebra/Determinat/Adjugate.agda:
```diff
@@ -0,0 +1,1943 @@
+{-# OPTIONS --safe #-}
```
remove

mortberg commented on Cubical/Algebra/Determinat/Adjugate.agda:
```diff
@@ -0,0 +1,1943 @@
+
+    deltaPropEq : (k l : Fin n) → k ≡ l → δ k l ≡ 1r
+    deltaPropEq k l e =
+      δ k l
+      ≡⟨ cong (λ a → δ a l) e ⟩
+      δ l l
+      ≡⟨ cong (λ a → if a then 1r else 0r) (==Refl l) ⟩
+      1r
+      ∎
+
+    deltaComm : (k l : Fin n) → δ k l ≡ δ l k
+    deltaComm k l = cong (λ a → if a then 1r else 0r) (==Sym k l)
+
+  -- Definition of the cofactor matrix
+  cof : FinMatrix R n n → FinMatrix R n n
+  cof {suc n} M i j = (MF (toℕ i +ℕ toℕ j)) ·  det {n} (minor i j M)
+
+  opaque
+    -- Behavior of the cofactor matrix under transposition
+    cofTransp : (M : FinMatrix R n n) → (i j : Fin n) → cof (M ᵗ) i j ≡ cof M j i
+    cofTransp {suc n} M i j =
+      MF (toℕ i +ℕ toℕ j) ·  det (minor i j (M ᵗ))
+      ≡⟨ cong (λ a → MF (toℕ i +ℕ toℕ j) · a) (detTransp ((minor j i M ᵗ))) ⟩
+      (MF (toℕ i +ℕ toℕ j) · det (minor j i M))
+      ≡⟨
+        cong
+        (λ a → MF (a) · det (minor j i M)) (+ℕ-comm (toℕ i) (toℕ j)) ⟩
+      (MF (toℕ j +ℕ toℕ i) · det (minor j i M))
+      ∎
+
+  -- Definition of the adjugate matrix
+  adjugate : FinMatrix R n n → FinMatrix R n n
+  adjugate M i j = cof M j i
+
+  opaque
+    -- Behavior of the adjugate matrix under transposition
+    adjugateTransp : (M : FinMatrix R n n) → (i j : Fin n) → adjugate (M ᵗ) i j ≡ adjugate M j i
+    adjugateTransp M i j = cofTransp M j i
+
+    -- Properties of WeakenFin
```
Move to theory for Fin

mortberg commented on Cubical/Algebra/Determinat/Adjugate.agda:
```diff
@@ -0,0 +1,1943 @@
+
+    deltaComm : (k l : Fin n) → δ k l ≡ δ l k
+    deltaComm k l = cong (λ a → if a then 1r else 0r) (==Sym k l)
+
+  -- Definition of the cofactor matrix
+  cof : FinMatrix R n n → FinMatrix R n n
+  cof {suc n} M i j = (MF (toℕ i +ℕ toℕ j)) ·  det {n} (minor i j M)
+
+  opaque
+    -- Behavior of the cofactor matrix under transposition
+    cofTransp : (M : FinMatrix R n n) → (i j : Fin n) → cof (M ᵗ) i j ≡ cof M j i
+    cofTransp {suc n} M i j =
+      MF (toℕ i +ℕ toℕ j) ·  det (minor i j (M ᵗ))
+      ≡⟨ cong (λ a → MF (toℕ i +ℕ toℕ j) · a) (detTransp ((minor j i M ᵗ))) ⟩
+      (MF (toℕ i +ℕ toℕ j) · det (minor j i M))
+      ≡⟨
+        cong
+        (λ a → MF (a) · det (minor j i M)) (+ℕ-comm (toℕ i) (toℕ j)) ⟩
+      (MF (toℕ j +ℕ toℕ i) · det (minor j i M))
+      ∎
+
+  -- Definition of the adjugate matrix
+  adjugate : FinMatrix R n n → FinMatrix R n n
+  adjugate M i j = cof M j i
+
+  opaque
+    -- Behavior of the adjugate matrix under transposition
+    adjugateTransp : (M : FinMatrix R n n) → (i j : Fin n) → adjugate (M ᵗ) i j ≡ adjugate M j i
+    adjugateTransp M i j = cofTransp M j i
+
+    -- Properties of WeakenFin
+    weakenPredFinLt : (k l : Fin (suc (suc n))) → toℕ k <' toℕ l → k ≤'Fin weakenFin (predFin l)
+    weakenPredFinLt {zero} zero one (s≤s z≤) = z≤
+    weakenPredFinLt {zero} one one (s≤s ())
+    weakenPredFinLt {zero} one (suc (suc ())) (s≤s le)
+    weakenPredFinLt {suc n} zero one (s≤s z≤) = z≤
+    weakenPredFinLt {suc n} zero (suc (suc l)) (s≤s le) = z≤
+    weakenPredFinLt {suc n} (suc k) (suc (suc l)) (s≤s (s≤s le)) = s≤s ( weakenPredFinLt {n} k (suc l) (s≤s le))
+
+    sucPredFin : (k l : Fin (suc (suc n))) → toℕ k <' toℕ l → suc (predFin l) ≡ l
```
move

mortberg commented on Cubical/Algebra/Determinat/Adjugate.agda:
```diff
@@ -0,0 +1,1943 @@
+               (λ a → a · (ind> (toℕ i) (toℕ j) ·
+               (M l i · (MF (toℕ k) · MF (toℕ i)) ·
+                (MF (toℕ (predFin l)) · MF (toℕ j) · M l (weakenFin j) ·
+                 det (minor (predFin l) j (minor k i M))))))
+               (+InvR 1r)
+             ⟩
+            (0r ·
+              (ind> (toℕ i) (toℕ j) ·
+               (M l i · (MF (toℕ k) · MF (toℕ i)) ·
+                (MF (toℕ (predFin l)) · MF (toℕ j) · M l (weakenFin j) ·
+                 det (minor (predFin l) j (minor k i M))))))
+            ≡⟨ RingTheory.0LeftAnnihilates (CommRing→Ring P') (ind> (toℕ i) (toℕ j) ·
+                                                                (M l i · (MF (toℕ k) · MF (toℕ i)) ·
+                                                                 (MF (toℕ (predFin l)) · MF (toℕ j) · M l (weakenFin j) ·
+                                                                  det (minor (predFin l) j (minor k i M))))) ⟩
+            0r
+            ∎)
+        ⟩
+       ∑ (λ (i : Fin (suc (suc n))) → ∑ (λ (j : Fin (suc n)) → 0r))
+       ≡⟨ ∑Compat
+         (λ (i : Fin (suc (suc n))) → ∑ (λ (j : Fin (suc n)) → 0r))
+         (λ (i : Fin (suc (suc n))) → 0r)
+         (λ (i : Fin (suc (suc n))) → ∑Zero {suc n} (λ (i : Fin  (suc n)) → 0r) (λ (i : Fin  (suc n)) → refl)) ⟩
+       ∑ (λ  (i : Fin (suc (suc n))) → 0r)
+       ≡⟨ ∑Zero (λ (i : Fin (suc (suc n))) → 0r) (λ  (i : Fin (suc (suc n))) → refl) ⟩
+       0r
+       ∎
+
+    adjugateInvRGcomponent : (M : FinMatrix R n n) → (k l : Fin n) → toℕ l <' toℕ k → (M ⋆ adjugate M) k l ≡  (det M ∘ 𝟙) k l
+    adjugateInvRGcomponent {suc n} M k l le =
+      ∑ (λ i → M k i · (MF(toℕ l +ℕ toℕ i) · det(minor l i M)) )
+      ≡⟨ adjugatePropRG M l k le  ⟩
+      0r
+      ≡⟨ sym (RingTheory.0RightAnnihilates (CommRing→Ring P') (det M)) ⟩
+      det M · 0r
+      ≡⟨ cong (λ a → det M · a) (sym (deltaPropSym k l le)) ⟩
+      (det M ∘ 𝟙) k l
+      ∎
+
+  strongenFin : {j : Fin (suc n)} → (i : Fin (suc n)) → toℕ i <' toℕ j → Fin n
```
move

mortberg replied:
with lemmas about it

mortberg replied:
Also, "strongen" should probably be "strengthen"?

mortberg commented on Cubical/Algebra/Determinat/Adjugate.agda:
```diff
@@ -0,0 +1,1943 @@
+      (det M ∘ 𝟙) k l
+      ∎
+
+  FinCompare : {n : ℕ} → (k l : Fin n) →  (k ≡ l) ⊎ ((toℕ k <' toℕ l) ⊎ (toℕ l <' toℕ k))
```
move

mortberg commented on Cubical/Algebra/Determinat/Base.agda:
```diff
@@ -0,0 +1,2504 @@
+{-# OPTIONS --safe #-}
```
remove

mortberg commented on Cubical/Algebra/Determinat/Base.agda:
```diff
@@ -0,0 +1,2504 @@
+            · det (minor k i (minor zero (weakenFin j) M)))
+         ∎)
+
+  weakenweakenFinLe : {n : ℕ} → (i : Fin (suc n)) → (j : Fin (suc n)) → toℕ i ≤' toℕ j → weakenFin i ≤'Fin weakenFin j
```
move

mortberg commented on Cubical/Algebra/Determinat/Base.agda:
```diff
@@ -0,0 +1,2504 @@
+             ind> (suc (toℕ j)) (toℕ i) ·
+               (a ·
+               M (suc k) (weakenFin i)
+               · M (weakenFin zero) (suc j)
+               · det (minor k i (minor zero (suc j) M))))
+           (DetRowAux6bAux k i j)
+         ⟩
+        (ind> (suc (toℕ j)) (toℕ i) ·
+          (MF (toℕ (suc k) +ℕ toℕ (suc i) +ℕ toℕ (suc j)) ·
+           M (suc k) (weakenFin i)
+           · M (weakenFin zero) (suc j)
+           · det (minor k i (minor zero (suc j) M))))
+        ∎)
+        ⟩
+       ∑
+         (λ j →
+            ∑
+            (λ i →
+               ind> (suc (toℕ j)) (toℕ i) ·
+               (MF (toℕ (suc k) +ℕ toℕ (suc i) +ℕ toℕ (suc j)) ·
+                M (suc k) (weakenFin i)
+                · M (weakenFin zero) (suc j)
+                · det (minor k i (minor zero (suc j) M)))))
+       ≡⟨ sym (∑∑ShiftSuc λ j i → (MF (toℕ (suc k) +ℕ toℕ (suc i) +ℕ toℕ j) ·
+                M (suc k) (weakenFin i)
+                · M (weakenFin zero) j
+                · det (minor k i (minor zero j M)))) ⟩
+       ∑
+         (λ j →
+            ∑
+            (λ i →
+               ind> (toℕ j) (toℕ i) ·
+               (MF (toℕ (suc k) +ℕ toℕ (suc i) +ℕ toℕ j) ·
+                M (suc k) (weakenFin i)
+                · M zero j
+                · det (minor k i (minor zero j M)))))
+       ∎
+
+  opaque
+    weakenFinLe : {n : ℕ} → (i : Fin (suc (suc n))) → (j : Fin (suc n)) → toℕ i ≤' toℕ j → i ≤'Fin weakenFin j
```
move

mortberg commented on Cubical/Algebra/Determinat/Minor.agda:
```diff
@@ -0,0 +1,193 @@
+{-# OPTIONS --safe #-}
```
remove

mortberg commented on Cubical/Algebra/Determinat/Minor.agda:
```diff
@@ -0,0 +1,193 @@
+open import Cubical.Foundations.Structure using (⟨_⟩)
+open import Cubical.Data.FinData
+open import Cubical.Data.FinData.Order using (_<'Fin_; _≤'Fin_)
+open import Cubical.Algebra.Ring
+open import Cubical.Algebra.Ring.Base
+open import Cubical.Algebra.Ring.BigOps
+open import Cubical.Algebra.Monoid.BigOp
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.CommRing.Base
+open import Cubical.Data.Nat.Order
+open import Cubical.Tactics.CommRingSolver
+
+module Minor (ℓ : Level) where
+
+  -- definition and properties to remove one Index, i.e. (removeIndex i) is the monoton ebbeding from {0,...,n-1} to {0,...,n} ommiting i.
+  removeIndex : {n : ℕ} → Fin (suc n) → Fin n → Fin (suc n)
+  removeIndex zero k = suc k
+  removeIndex (suc i) zero = zero
+  removeIndex (suc i) (suc k) = suc (removeIndex i k)
+
+  -- On {0,...,i-1} the map (removeIndex i) is the identity.
+  removeIndexId : {n : ℕ} → (i : Fin (suc n)) → (k : Fin n) → (suc k) ≤'Fin i → (removeIndex i k) ≡ weakenFin k
+  removeIndexId (suc i) zero le = refl
+  removeIndexId (suc i) (suc k) (s≤s le) =  cong (λ a → suc a) (removeIndexId i k le)
+
+  -- On {i,...,n-1} the map (removeIndex i) is the successor function.
+  removeIndexSuc : {n : ℕ} → (i : Fin (suc n)) → (k : Fin n) → i ≤'Fin weakenFin k  → (removeIndex i k) ≡ suc k
+  removeIndexSuc zero k le = refl
+  removeIndexSuc (suc i) (suc k) (s≤s le) = cong (λ a → suc a) (removeIndexSuc i k le)
+
+  -- A formula for commuting removeIndex maps
+  removeIndexComm :  {n : ℕ} → (i j : Fin (suc n)) → (k : Fin n) → i ≤'Fin j →
+    removeIndex (suc j) (removeIndex i k) ≡ removeIndex (weakenFin i) (removeIndex j k)
+  removeIndexComm zero j k le = refl
+  removeIndexComm (suc i) (suc j) zero le = refl
+  removeIndexComm (suc i) (suc j) (suc k) (s≤s le) =  cong (λ a → suc a) (removeIndexComm i j k le)
+
+  -- remove the j-th column of a matrix
+
+  remove-column : {A : Type ℓ} {n m : ℕ} (j : Fin (suc m)) (M : FinMatrix A n (suc m)) → FinMatrix A n m
```
Rename to removeCol

mortberg commented on Cubical/Algebra/Determinat/Minor.agda:
```diff
@@ -0,0 +1,193 @@
+    ∎
+
+  -- remove the i-th row of a matrix
+  remove-row : {A : Type ℓ} {n m : ℕ} (i : Fin (suc n)) (M : FinMatrix A (suc n) m) → FinMatrix A n m
```
Rename to removeRow

mortberg commented on Cubical/Algebra/Determinat/Minor.agda:
```diff
@@ -0,0 +1,193 @@
+    ∎
+
+  -- remove the i-th row of a matrix
+  remove-row : {A : Type ℓ} {n m : ℕ} (i : Fin (suc n)) (M : FinMatrix A (suc n) m) → FinMatrix A n m
+  remove-row i M k l = M (removeIndex i k) l
+
+  remove-row-comm : {A : Type ℓ} {n m : ℕ} (i : Fin (suc n)) (j : Fin (suc n)) (M : FinMatrix A (suc (suc n)) m) (k : Fin n) (l : Fin  m) →
+    j ≤'Fin i →
+    remove-row j (remove-row (suc i) M) k l ≡  remove-row i (remove-row (weakenFin j) M) k l
+  remove-row-comm i j M k l le =
+    remove-row j (remove-row (suc i) M) k l
+    ≡⟨ refl ⟩
+      M (removeIndex (suc i) (removeIndex j k)) l
+    ≡⟨ cong (λ a → M a l) ( removeIndexComm j i k le) ⟩
+    M (removeIndex (weakenFin j) (removeIndex i k)) l
+    ≡⟨ refl ⟩
+     remove-row i (remove-row (weakenFin j) M) k l
+    ∎
+
+  remove-column-row-comm : {A : Type ℓ} {n m : ℕ}  (i : Fin (suc n)) (j : Fin (suc m)) (M : FinMatrix A (suc n) (suc m)) (k : Fin n) (l : Fin m) →  (remove-row i (remove-column j M)) k l ≡  (remove-column j (remove-row i M)) k l
```
Long line. There are other examples in this file as well

mortberg commented on Cubical/Algebra/Determinat/RingSum.agda:
```diff
@@ -0,0 +1,349 @@
+{-# OPTIONS --cubical #-}
```
remove

mortberg commented on Cubical/Algebra/Determinat/RingSum.agda:
```diff
@@ -0,0 +1,349 @@
+{-# OPTIONS --cubical #-}
+{-# OPTIONS --safe #-}
```
remove

mortberg reviewed (changes requested):
Many thanks! Lots of good stuff! I had some suggestions for improvements and then it can be merged

FranziskusWiesnet commented:
Thank you for the review. I have now addressed the review comments and updated the branch to the current master.
The files I modified as part of the determinant development and review fixes are:

- The four files in Cubical/Algebra/Determinat/
- Cubical/Data/FinData/Properties.agda
- Cubical/Data/FinData/Order.agda

Since I based the revised version on the current upstream master, the PR also contains many unrelated changes that have been added to master since the original submission.

mortberg commented:
I think this is ready to be merged now, except for one typo: The directory should be `Algebra/Determinant` not `Algebra/Determinat` (note the missing `n`)

FranziskusWiesnet commented:
Thank you! The typo is fixed.

mortberg commented:
Perfect! Will merge once the CI is finished
